runtime/compiler/optimizer/EscapeAnalysis.cpp

```diff
@@ -2178,69 +2231,134 @@ bool TR_EscapeAnalysis::checkOtherDefsOfLoopAllocation(TR::Node *useNode, Candid
+                      _aliasesOfOtherAllocNode->get(defNode2->getSymbolReference()->getReferenceNumber()))
                      {
                      rhsIsHarmless = true;
                      break;
                      }
                   }
+               }
 
-               if (!rhsIsHarmless)
+            if (!rhsIsHarmless)
+               {
+               // Another special case when it is certain that the rhs of the def is not a candidate allocation from a prior iteration
+               // In this case we are loading a value from an Integer cache anyway and that should be an allocation that has already escaped
+               // that has nothing to do with the candidate allocation
+               //
+               if (firstChild->getOpCode().hasSymbolReference() &&
+                   firstChild->getSymbol()->isArrayShadowSymbol())
                   {
-                  if (firstChild->getOpCode().hasSymbolReference() &&
-                      firstChild->getSymbol()->isArrayShadowSymbol())
+                  TR::Node *addr = firstChild->getFirstChild();
+                  if (addr->getOpCode().isArrayRef())
                      {
-                     TR::Node *addr = firstChild->getFirstChild();
-                     if (addr->getOpCode().isArrayRef())
-                        {
-                        TR::Node *underlyingArray = addr->getFirstChild();
-
-                        int32_t fieldNameLen = -1;
-                        char *fieldName = NULL;
-                        if (underlyingArray && underlyingArray->getOpCode().hasSymbolReference() &&
-                            underlyingArray->getSymbolReference()->getSymbol()->isStaticField())
-                           {
-                           fieldName = underlyingArray->getSymbolReference()->getOwningMethod(comp())->staticName(underlyingArray->getSymbolReference()->getCPIndex(), fieldNameLen, comp()->trMemory());
-                           }
+                     TR::Node *underlyingArray = addr->getFirstChild();
 
-                        if (fieldName && (fieldNameLen > 0) &&
-                           !strncmp(fieldName, "java/lang/Integer$IntegerCache.cache", 36))
-                           rhsIsHarmless = true;
+                     int32_t fieldNameLen = -1;
+                     char *fieldName = NULL;
+                     if (underlyingArray && underlyingArray->getOpCode().hasSymbolReference() &&
+                         underlyingArray->getSymbolReference()->getSymbol()->isStaticField())
+                        {
+                        fieldName = underlyingArray->getSymbolReference()->getOwningMethod(comp())->staticName(underlyingArray->getSymbolReference()->getCPIndex(), fieldNameLen, comp()->trMemory());
                         }
+
+                     if (fieldName && (fieldNameLen > 0) &&
+                         !strncmp(fieldName, "java/lang/Integer$IntegerCache.cache", 36))
```

Comment:
does long have an equivalent cache that should be treated the same way for symmetry?

Reply:
It seems it does, as do `java.lang.Byte` and `java.lang.Character`

Reply:
Expanded the set of caches checked for in commit 92b0cb068aa2ec511d8df1d4cc8a071eac10ca4d